Fix can not remove Group Plan tag

nikonhub opened the pull request:
- Allow to change name and remove a group tag
- Add a rule to check if the group tag is an active one to hide the
trash icon

[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #11522

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
![image](https://user-images.githubusercontent.com/45662151/69009221-aeee2680-0953-11ea-95e2-6c3670a06119.png)

Client : 
- Allow to edit names of all tags
- Allow to remove all tags except if the user belongs to it. (guild or party group plan subscription)

Server :
- api/v4/user update tags. Change rules of tags that will be kept. Do not by default keep all group tags but check if it belongs to an active user guild or party instead.
- When a group leader assigns a new task to a member, do not reset the tag name of this member.


[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 07bd9063-362f-40c6-8d0c-1fa8b977243d

paglias commented on website/server/libs/user/index.js:
```diff
@@ -126,39 +126,34 @@ export async function update (req, res, { isV3 = false }) {
-        const oldI = removedTagsIds.findIndex(id => id === t.id);
-        if (oldI > -1) {
-          removedTagsIds.splice(oldI, 1);
+        if (user.party._id && user.party._id === tag.group) {
```
this seems to be only checking the user's party but you can also be part of a group that is not your party (see `user.groups`)

nikonhub replied:
Do you mean user.guilds ? If yes I check it just after at line 140.

paglias replied:
sorry I missed that

paglias reviewed (changes requested):
left one comment but this is also going to need some integration / unit tests to make sure the API is working correctly, thanks @Nikosmonaut !

nikonhub commented:
> this is also going to need some integration / unit tests to make sure the API is 

I added 2 integrations tests. No sure about about unit ones though.

paglias reviewed (changes requested):
@Nikosmonaut sorry for the delay. I gave a better look at the PR and I think it's a bit overly complicated. 

Instead of changing the way the tags are updated I think it would be better to remove the `group` attribute from your tag when you leave or are removed from a group, like the original issue said

```
The correct behaviour is that the tag should become one of your personal tags. I.e., the group attribute should be removed from the tag (see the sample tag below which shows that attribute).
```

This way the changes to the client side won't be necessary. The code executed when you leave a group is here https://github.com/HabitRPG/habitica/blob/develop/website/server/models/group.js#L1315

nikonhub commented:
Hello.

No problem for the delay. (But yes it kind of gets old in my head the reasons for this solution).
Like I said [here](https://github.com/HabitRPG/habitica/issues/11522#issuecomment-552503278) we may face another unwanted behavior if the user is reinvited twice to the same group. Because he's "personal" old group tag would be reused.

```
    "tags": [
        {
            "id": "d2397cb7-e92d-42dc-aba8-6686cab105e3",
            "name": "The Groupish Group",
            "group": "d2397cb7-e92d-42dc-aba8-6686cab105e3"
        },
```
I think the perfect solution would be to not have the same id. Tag id !== group. So we could just delete the group key and we're done. But it would need to change the whole database :/

If you're ok with this side effect of reusing the same tag, I revert my changes to just remove the group key.

Edit : 
It is this code I'm talking about.
Maybe another solution would be to findIndex by { group: group._id }. From now have id !== group and keep old tags as is. But I it seems confusing.

```
  const userTags = user.tags;
  const i = _.findIndex(userTags, { id: group._id });

  if (i === -1) {
    userTags.push({
      id: group._id,
      name: group.name,
      group: group._id,
    });
  }
```

paglias commented:
> I think the perfect solution would be to not have the same id. Tag id !== group. So we could just delete the group key and we're done. But it would need to change the whole database :/

> If you're ok with this side effect of reusing the same tag, I revert my changes to just remove the group key.

Good point. I'd say I'm okay with reusing the same id but let's see if @Alys is okay with that. If you don't hear anything in the next few days feel free to reuse the id

Alys commented:
@Nikosmonaut I'm okay with reusing the same id too. Thanks for thinking through the options and implications!

Commit:
Unlink group tags when a user leaves it

paglias commented:
Thanks @Nikosmonaut and sorry for the delay! The code fix looks good but it would need some tests, if you want to add them great but if you have any problem let me know and I can add them :)

Commit:
Add test for leaving group

After leaving group the previous associated tag should no more be bound
to this group.

nikonhub commented:
Hello @paglias.

Just added a test.
Sorry for the delay. Christmas holidays not obvious to find time.

paglias commented:
Hi @Nikosmonaut  , sorry for the delay but I'm still catching up from the end of year holidays, I hope to get to this PR in the next couple of days

paglias commented:
It took a while but I finally managed to merge this, thanks @Nikosmonaut ! I've notated this PR towards your 3rd tier
